cmake/modules/FindFlatBuffers.cmake

```diff
@@ -16,24 +16,63 @@
+# FLATBUFFERS_FLATC_EXECUTABLE - the flatc compiler executable
+# FLATBUFFERS_INCLUDE_DIRS - the flatbuffers include directory
 #
 # Provides:
-# * FLATBUFFERS_GENERATE_C_HEADERS(Name <files>) creates the C++ headers
+#
+# FLATBUFFERS_GENERATE_C_HEADERS(Name <files>) creates the C++ headers
 #   for the given flatbuffer schema files.
 #   Returns the header files in ${Name}_OUTPUTS
 
-set(FLATBUFFERS_CMAKE_DIR ${CMAKE_CURRENT_LIST_DIR})
+if(ENABLE_INTERNAL_FLATBUFFERS)
+  include(ExternalProject)
+  file(STRINGS ${CMAKE_SOURCE_DIR}/tools/depends/native/flatbuffers-native/Makefile VER REGEX "^[ ]*VERSION[ ]*=.+$")
+  string(REGEX MATCH "VERSION=[^ ]*" FLATBUFFERS_VER "${VER}")
+
+  # allow user to override the download URL with a local tarball
+  # needed for offline build envs
+  if(FLATBUFFERS_URL)
+    get_filename_component(FLATBUFFERS_URL "${FLATBUFFERS_URL}" ABSOLUTE)
+  else()
+    #set(FLATBUFFERS_URL http://mirrors.kodi.tv/build-deps/sources/flatbuffers-${FLATBUFFERS_VER}.tar.gz)
+    set(FLATBUFFERS_URL https://github.com/google/flatbuffers/archive/v${FLATBUFFERS_VER}.tar.gz)
+  endif()
+  if(VERBOSE)
+    message(STATUS "FLATBUFFERS_URL: ${FLATBUFFERS_URL}")
+  endif()
 
-find_program(FLATBUFFERS_FLATC_EXECUTABLE NAMES flatc)
-find_path(FLATBUFFERS_INCLUDE_DIR NAMES flatbuffers/flatbuffers.h)
+  set(FLATBUFFERS_FLATC_EXECUTABLE ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/bin/flatc)
+  set(FLATBUFFERS_INCLUDE_DIR ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/include)
+
+  externalproject_add(flatbuffers
+                      URL ${FLATBUFFERS_URL}
+                      DOWNLOAD_DIR ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/download
+                      PREFIX ${CORE_BUILD_DIR}/flatbuffers
+                      CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}
+                                 -DCMAKE_BUILD_TYPE=Release
+                                 -DFLATBUFFERS_CODE_COVERAGE=OFF
+                                 -DFLATBUFFERS_BUILD_TESTS=OFF
+                                 -DFLATBUFFERS_INSTALL=ON
+                                 -DFLATBUFFERS_BUILD_FLATLIB=OFF
+                                 -DFLATBUFFERS_BUILD_FLATC=ON
+                                 -DFLATBUFFERS_BUILD_FLATHASH=OFF
+                                 -DFLATBUFFERS_BUILD_GRPCTEST=OFF
+                                 -DFLATBUFFERS_BUILD_SHAREDLIB=OFF
+                                 "${EXTRA_ARGS}"
+                      BUILD_BYPRODUCTS ${FLATBUFFERS_FLATC_EXECUTABLE})
+  set_target_properties(flatbuffers PROPERTIES FOLDER "External Projects")
+else()
+  find_program(FLATBUFFERS_FLATC_EXECUTABLE NAMES flatc)
+  find_path(FLATBUFFERS_INCLUDE_DIR NAMES flatbuffers/flatbuffers.h)
+endif()
 
 include(FindPackageHandleStandardArgs)
 find_package_handle_standard_args(flatbuffers
   DEFAULT_MSG FLATBUFFERS_FLATC_EXECUTABLE FLATBUFFERS_INCLUDE_DIR)
 
-if(FLATBUFFERS_FOUND)
   function(FLATBUFFERS_GENERATE_C_HEADERS Name)
```

Comment:
@Rechi This macro uses `${FLATBUFFERS_FLATC_EXECUTABLE}`, which is defined in this cmake module, but the variable is empty when the macro is run. How can I propagate its value to this macro? It worked before I enabled building internal.

Reply:
it would work if you made it a macro, functions have locally scoped vars.
I'd move the function to Macros.cmake though.